drivers/timer/nrf_rtc_timer.c

```diff
@@ -395,9 +395,13 @@ uint32_t sys_clock_elapsed(void)
 
 uint32_t sys_clock_cycle_get_32(void)
 {
-	k_spinlock_key_t key = k_spin_lock(&lock);
-	uint32_t ret = counter_sub(counter(), last_count) + last_count;
+	volatile uint32_t *base = &last_count;
+	uint32_t ret, anchor;
+
+	do {
+		anchor = *base;
+		ret = counter_sub(counter(), anchor) + anchor;
+	} while (anchor != *base);
```

Comment:
I'm not clear on what the loop is giving you here.  Why not just:

    uint32_t last = last_count;
    
    return counter_sub(counter(), last) + last;

This should return a valid timestamp even if `last_count` is being updated simultaneously.   To the caller it's equivalent to the loop when the update happens right after the loop exit condition is met.

Reply:
I agree with @pabigot. There's no need for a loop here.